sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ResolveDefaultColumns.scala

```diff
@@ -18,14 +18,227 @@
+import org.apache.spark.sql.catalyst.rules.Rule
+import org.apache.spark.sql.catalyst.util.ResolveDefaultColumns._
 import org.apache.spark.sql.internal.SQLConf
 import org.apache.spark.sql.types._
 
+/**
+ * This is a rule to process DEFAULT columns in statements such as CREATE/REPLACE TABLE.
+ *
+ * Background: CREATE TABLE and ALTER TABLE invocations support setting column default values for
+ * later operations. Following INSERT, and INSERT MERGE commands may then reference the value
+ * using the DEFAULT keyword as needed.
+ *
+ * Example:
+ * CREATE TABLE T(a INT DEFAULT 4, b INT NOT NULL DEFAULT 5);
+ * INSERT INTO T VALUES (1, 2);
+ * INSERT INTO T VALUES (1, DEFAULT);
+ * INSERT INTO T VALUES (DEFAULT, 6);
+ * SELECT * FROM T;
+ * (1, 2)
+ * (1, 5)
+ * (4, 6)
+ *
+ * @param analyzer analyzer to use for processing DEFAULT values stored as text.
+ * @param catalog the catalog to use for looking up the schema of INSERT INTO table objects.
+ * @param insert the enclosing INSERT statement for which this rule is processing the query, if any.
+ */
+case class ResolveDefaultColumns(
+  analyzer: Analyzer,
+  catalog: SessionCatalog,
+  insert: Option[InsertIntoStatement] = None) extends Rule[LogicalPlan] {
+  override def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsWithPruning(
+    (_ => SQLConf.get.enableDefaultColumns), ruleId) {
+    case i@InsertIntoStatement(_, _, _, _, _, _)
+      if i.query.collectFirst { case u: UnresolvedInlineTable => u }.isDefined =>
+      // Create a helper instance of this same rule with the `insert` argument populated as `i`.
+      // Then recursively apply it on the `query` of `i` to transform the result. This recursive
+      // application lets the below case matching against `UnresolvedInlineTable` to trigger
+      // anywhere that operator may appear in the descendants of `i.query`.
+      val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
+      val newQuery = helper.apply(i.query)
+      i.copy(query = newQuery)
+
+    case table: UnresolvedInlineTable
+      if insert.isDefined &&
+        table.rows.nonEmpty && table.rows.forall(_.size == table.rows(0).size) =>
+      val expanded: UnresolvedInlineTable = addMissingDefaultColumnValues(table).getOrElse(table)
+      replaceExplicitDefaultColumnValues(analyzer, expanded).getOrElse(table)
+
+    case i@InsertIntoStatement(_, _, _, project: Project, _, _)
+      if !project.resolved =>
+      val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
+      val replaced: Option[Project] = helper.replaceExplicitDefaultColumnValues(analyzer, project)
+      if (replaced.isDefined) i.copy(query = replaced.get) else i
+
+    case i@InsertIntoStatement(_, _, _, project: Project, _, _)
+      if project.resolved =>
+      val helper = ResolveDefaultColumns(analyzer, catalog, Some(i))
+      val expanded: Project = helper.addMissingDefaultColumnValues(project).getOrElse(project)
```

Comment:
Could you explain why we only need to expand the project when it is resolved?

Reply:
I removed this check. It turns out this was a hold-out from the previous version of the rule logic. We do not need to keep this anymore, since the "check analysis" step that errors-out in response to the unresolved DEFAULT column happens later in the sequence. The code is simpler now.